### DISCUSSION TITLE
Overall database architecture

### DISCUSSION THREAD

**Discussion by LunaUrsa** [TripSit/drugs#20] [Development]
Moonbear realizes that this github setup is not ideal, and is open to ideas that can satisfy the following requirements:

1. Enable applications to pull the database
- The tripsit API will utilize this JSON file so a full API isnt necessary, just a way to download the JSON file to the application.
2. Allow public users to suggest changes.
3. A process to review and accept those changes.
4. A way for the public to review the history of changes.

Wiki does 2, 3, and 4, but without 1, the API won't be able to use the most recent version.
Postgres does 1, but we would need to build something custom for 2, 3 and 4.
GitHub is the only solution that does all four, so we created this separate repository to track all changes to the drug information.

**Comment by LunaUrsa** [comment]
Talking with Pfish on the discord, we're thinking if we were to do something custom:

- Move JSON to Postgres
- Export the Postgres as a series of CSV
- Upload the CSV to github
- Modify the Factsheets web page:
    - Allow users to login with the TripSit auth system
    - Allow users to modify the page and submit changes
    - Submission would create a github issue to modify the CSV backend
- Modifications to the CSV files would trigger a deployment to postgres
- Rest API would be configured to return a JSON-y response after querying the postgres

**Comment by mastfissh** [comment]
So, I absolutely love postgres. I've used it professionally for more than a decade. It's the best database technology that currently exists,  and it's fantastic at what it does. However, for what you want to do, there are some very significant downsides.

Postgres requires a separate database server. That server takes a significant amount of maintenance (or paying amazon etc to manage it). Synchronizing changes between different postgres server instances (to allow for users to  suggest changes, wiki style) can be very challenging. Anyone who wants to tinker with the application will also now need their own postgres server - this is a pretty big obstacle. They will also need to regularly synchronize their DB. None of these are insurmountable problems, but the effort involved is very significant. 

If we want the source of truth to be inside git (like the CSV files), I think SQLite might be a better option? It's not as fancy as postgres, but it doesn't require a separate server, so it's much, much easier to work with operationally.

**Comment by utaninja** [comment]
Is this updated and still needed now?

**Reply by LunaUrsa** [reply]
@ELLEation has been working on something, last update was that they were working on API stuff
